dev-docs/git-practices.md

```diff
@@ -0,0 +1,133 @@
+local copy of your fork with a single remote called `origin`. In this document,
+we will refer to `origin` as your developer fork, and `upstream` as `hail-is/hail`.
+You can check which origins you have configured by running `git remote -v`.
+
+While feature branches will live on your fork, those branches are still going to want
+to be based on the latest changes in `hail-is/hail:main`. So we will add the
+upstream repository as another remote so we can pull in those changes.
+
+```bash
+git remote add upstream https://github.com/hail-is/hail.git
+```
+
+If you run `git remote -v` again, you should see something like the following:
+
+```
+origin	https://github.com/<your-github-username>/hail.git (fetch)
+origin	https://github.com/<your-github-username>/hail.git (push)
+upstream	https://github.com/hail-is/hail.git (fetch)
+upstream	https://github.com/hail-is/hail.git (push)
+```
+
+To get the latest changes on your local `main` branch, run
+
+```bash
+git checkout main
+git pull upstream main
+```
+
+Now if you create a feature branch like `git checkout -b my-feature`, it will
+be based on the latest upstream changes.
+
+
+## While developing a feature
+
+`hail-is/hail:main` moves quickly, and it is likely that it will have progressed
+significantly while you work on a feature. This is not in itself a problem,
+but can cause headaches if changes in upstream `main` conflict with the code you
+are working on. To cope with this, we recommend [rebasing](https://git-scm.com/docs/git-rebase#_description)
+regularly and often. Rebasing will incoporate the new changes from main into your
+branch and allow you to resolve any conflicts that might have developed along the
+way.
+
+To rebase your feature branch on the latest upstream main, run
+
+```bash
+git checkout main
+git pull upstream main
+
+git checkout <feature_branch>
+git rebase -i main
+```
+
+The `-i` here is optional but extremely informative, and can help you understand
+what the rebase is doing. You can leave all your commits as `pick`, and keep
+all your commits while perfoming the rebase, or you can change some commits to
+`squash` to collapse your changes into a smaller number of commits. When opening
+your change for review, it can be helpful for the reviewers if you squash your
+branch into a small number of self-contained commits. For example, if your change
+requires upgrading a dependency, it is helpful to put the dependency upgrades
+into a separate commmit from the code changes.
+
+
+## Making a PR
+
+When a feature branch is ready for PR, push the branch to your fork:
+
+```bash
+git push origin <feature_branch>
+```
+
+If you have already done this before but have since rebased, you may get
+an error because the history on your fork is no longer a prefix of your local
+branch's history. Force push your branch to overwrite the history on GitHub
+with that of your local branch.
+
+```bash
+git push -f origin <feature_branch>
```

Comment:
I think we should advocate for --force-with-lease. Maybe even add a dev-bin like `alias gp="git push --force-with-lease" origin`

Reply:
Didn't know about this, awesome!